Batch verification benchcoin

Eunovo opened the pull request.

fjahr committed:
BatchVerify: Add basic schnorr sig batch verification in ConnectBlock()

> build/src/bench/bench_bitcoin -filter=ConnectBlock
|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      316,756,339.00 |                3.16 |    0.1% |  191,713,738.00 |   72,606,304.00 |  2.640 |  13,982,663.00 |    0.7% |      3.49 | `ConnectBlock`

- Add batchverify.cpp to cmake lists
- Make custom Batch type for secp256k1_batch
  this is a bit of a hack, we can probably be smarter using a named api struct for
  the batch verification module

Eunovo committed:
Benchmark Chainstate::ConnectBlock duration

Measure ConnectBlock performance of full taproot blocks without signature caching
> cmake -B build -DBUILD_BENCH=ON
> cmake --build build
> build/src/bench/bench_bitcoin -filter=ConnectBlock

|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      137,558,591.00 |                7.27 |    0.0% |1,447,910,640.00 |  468,414,690.00 |  3.091 |  35,813,577.00 |    1.7% |      1.52 | `ConnectBlock`

josibake committed:
validation: assumeutxo benchmarking patch

introduce two commandline options for assumeutxo, specifically for
benchmarking. these commands are:

- pausebackgroundsync - an option lets the user pause the verification
                        of historical blocks in the background
-loadutxosnapshot=<path> - load an assumeutxo snapshot on startup,
                           instead of needing to go through the rpc
                           command. the node will shutdown immediately
                           after the snapshot has been loaded

this path is not meant for general use and is instead just for making it
more ergonomic to use assumeutxo for benchmarking IBD changes. the
benefits of using assumeutxo here are we can start from an arbitrary
height and sync to chaintip to collect relevant data quickly. using
assumeutxo means we can make whatever changes we need to the
chainstatedb, since it will be created fresh from the snapshot.

note, to use the loadutxosnapshot option, you must first run:

    ./build/src/bitcoind -stopatheight=1

this makes the node do a header sync and then shut down. this is because
assumeutxo will not load a snapshot unless the base block is in the
header chain. we could remove this requirement, but this patch is meant
to be as minimal as possible, and this also allows us to perform heaeder
sync as a preparation commit for a benchmark, which helps keep IBD
benchmarks more focused on strictly measuring IBD.

next, run:

    ./build/src/bitcoind -loadutxosnapshot=<path>

the node will shutdown after the snapshot is loaded. finally, run:

    ./build/src/bitcoind -pausebackgroundsync=1

for the actual benchmarking step. this ensures only the sync to chaintip
is benchmarked and the load snapshot step is not included in the
measurement.

Co-authored-by: Sjors Provoost <10217+sjors@users.noreply.github.com>

Eunovo committed:
WIP: Assign one batch to each thread for parallel script checking

- Creating one batch per thread improves batch validation performance by 17%
- Each batch is used by all the checks done on each thread and destroy when the check queue is empty
- ConnectBlock Benchmark results:

|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      113,877,596.00 |                8.78 |    0.1% |1,174,644,321.00 |  387,677,040.00 |  3.030 |  32,164,312.00 |    2.0% |      1.26 | `ConnectBlock`

Eunovo committed:
Cache batch signature validation result

willcl-ark committed:
env: add shell.nix

willcl-ark committed:
env: add uv for python deps

willcl-ark committed:
env: add justfile command runner

Eunovo committed:
Set -DSECP256K1_EXPERIMENTAL=ON to fix ci build failure

SECP256K1_EXPERIMENTAL=ON is hardcoded in src/secp256k1/CMakeLists.txt:159 and I can build locally without the setting this option manually
But this doesn't work on the ci. I couldn't find the cause. I don't see any command overriding the default value
I added this option as a hack to fix the error

willcl-ark commented:
👀

willcl-ark commented:
I do seem to see this defaulting to `OFF` in master:

```fish
$ rg SECP256K1_EXPERIMENTAL
src/secp256k1/CMakeLists.txt
159:option(SECP256K1_EXPERIMENTAL "Allow experimental configuration options." OFF)
160:if(NOT SECP256K1_EXPERIMENTAL)
162:    message(FATAL_ERROR "ARM32 assembly is experimental. Use -DSECP256K1_EXPERIMENTAL=ON to allow.")
398:if(SECP256K1_EXPERIMENTAL)
```

Eunovo commented:
> I do seem to see this defaulting to `OFF` in master:

@willcl-ark 
I change this in https://github.com/bitcoin-dev-tools/benchcoin/pull/41/commits/55a2f7a8409b37d502d8b6d1928721a7b01fe6fd#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

Maybe my change is being overwritten?

Eunovo committed:
Set -assumevalid=0, VERIFY ALL!

This PR makes improvements to script verification in connectblock
Assumevalid disables script verification which prevents us from getting the benchmark results we are looking for

Eunovo committed:
Test with only one script verification thread

Eunovo committed:
Use batch for single thread script verification too

willcl-ark committed:
ci: add benchcoin workflows

Co-authored-by: David Gumberg <davidzgumberg@gmail.com>
Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

willcl-ark committed:
doc: update README.md

willcl-ark committed:
ci: remove bitcoin core ci

Eunovo committed:
Squashed 'src/secp256k1/' changes from 0cdc758a56..9827fbfd41

9827fbfd41 Setup batch experimental module on cmake
89a54e23de Clear warnings
fb52de16b9 WIP: Silent merge conflicts
f1a167edff batch: Generate graphs for batch verification speed up
63823d9423 batch, extrakeys: Add benchmark for batch verify and `tweak_add_check`
b2f217c0ec batch: Add tests for `batch_add_*` APIs
ac564e9a23 batch,ecmult: Add tests for core batch APIs and `strauss_batch` refactor
f0c9b8801d batch: Add API usage example
3ab4a4c81e batch: Add `batch_add_*` APIs
ee9c242a80 batch, ecmult: Add `batch_verify` API and refactor `strauss_batch`
a00fbe43aa batch: Add `create` and `destroy` APIs
e7e627fa52 batch: Initialize an experimental batch module

git-subtree-dir: src/secp256k1
git-subtree-split: 9827fbfd4164080b3de5e1238bc75fb4f07dd5e1

Eunovo committed:
Merge commit '7f6c435f4cbe81e8892698eab5c50a6ecfac5a1e' into secp-benchcoin
